src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3431,7 +3426,8 @@ private void CheckForStructBadInitializers(DeclaredMembersAndInitializersBuilder
 
             if (builder.RecordDeclarationWithParameters is not null)
             {
-                Debug.Assert(builder.RecordDeclarationWithParameters is RecordStructDeclarationSyntax { ParameterList: not null });
+                Debug.Assert(builder.RecordDeclarationWithParameters is RecordDeclarationSyntax { ParameterList: not null } record
+                    && record.Kind() == SyntaxKind.RecordStructDeclaration);
```

Comment:
This made me realize a `void Deconstruct(this CSharpSyntaxNode, out SyntaxKind)` might be actually useful. Currently matching multi-kind nodes is kind of ugly..

Reply:
That's a really interesting suggestion. I've also struggled with the fact that Kind is a method and can't be used with pattern matching.